grudge/trace_pair.py

```diff
@@ -275,58 +280,74 @@ def connected_ranks(dcoll: DiscretizationCollection):
 class _RankBoundaryCommunication:
     base_tag = 1273
 
-    def __init__(self, dcoll: DiscretizationCollection,
-                 remote_rank, vol_field, tag=None):
-        self.tag = self.base_tag
-        if tag is not None:
-            self.tag += tag
-
-        self.dcoll = dcoll
-        self.array_context = vol_field.array_context
-        self.remote_btag = BTAG_PARTITION(remote_rank)
-        self.bdry_discr = dcoll.discr_from_dd(self.remote_btag)
+    def __init__(self,
+                 dcoll: DiscretizationCollection,
+                 array_container: ArrayOrContainerT,
+                 remote_rank, tag=None):
+        actx = get_container_context_recursively(array_container)
+        btag = BTAG_PARTITION(remote_rank)
 
-        from grudge.op import project
+        local_bdry_data = project(dcoll, "vol", btag, array_container)
+        comm = dcoll.mpi_communicator
 
-        self.local_dof_array = project(dcoll, "vol", self.remote_btag, vol_field)
+        self.dcoll = dcoll
+        self.array_context = actx
+        self.remote_btag = btag
+        self.bdry_discr = dcoll.discr_from_dd(btag)
+        self.local_bdry_data = local_bdry_data
+        self.local_bdry_data_np = \
+            to_numpy(flatten(self.local_bdry_data, actx), actx)
 
-        local_data = self.array_context.to_numpy(flatten(self.local_dof_array))
-        comm = self.dcoll.mpi_communicator
+        self.tag = self.base_tag
+        if tag is not None:
+            self.tag += tag
 
-        self.send_req = comm.Isend(local_data, remote_rank, tag=self.tag)
-        self.remote_data_host = np.empty_like(local_data)
-        self.recv_req = comm.Irecv(self.remote_data_host, remote_rank, self.tag)
+        # Here, we initialize both send and recieve operations through
+        # mpi4py `Request` (MPI_Request) instances for comm.Isend (MPI_Isend)
+        # and comm.Irecv (MPI_Irecv) respectively. These initiate non-blocking
+        # point-to-point communication requests and require explicit management
+        # via the use of wait (MPI_Wait, MPI_Waitall, MPI_Waitany, MPI_Waitsome),
+        # test (MPI_Test, MPI_Testall, MPI_Testany, MPI_Testsome), and cancel
+        # (MPI_Cancel). The rank-local data `self.local_bdry_data_np` will have its
+        # associated memory buffer sent across connected ranks and must not be
+        # modified at the Python level during this process. Completion of the
+        # requests are handled in :meth:`finish`.
+        #
+        # For more details on the mpi4py semantics, see:
+        # https://mpi4py.readthedocs.io/en/stable/overview.html#nonblocking-communications
+        #
+        # NOTE: mpi4py does hold onto the memory buffer for `self.local_bdry_data_np`
+        # until the send/recv requests are complete.
```

Comment:
```suggestion
        # NOTE: mpi4py currently (2021-11-03) holds a reference to the send memory buffer for
        # (i.e. `self.local_bdry_data_np`) until the send requests is complete, however it is not clear
        # that this is documented behavior. We hold on to the buffer (via the instance attribute)
        # as well, just in case.
```